File: apps/ios/Packages/StoryArcKit/Sources/LibraryFeature/LibraryStates.swift
public import SwiftUI

internal import DesignSystem
public import StoryArcCore

// What the library shows before it shows publications: a walk in progress, a summary of
// what it skipped, and the two empty states — nothing added, and something added that has
// nothing on it. Split out of `LibraryView` for the same reason as the controls.
//
// It used to end with a list of configured sources and their connection dots. That is
// configuration, and §6.2 of the design direction puts configuration in Settings and
// nowhere else on the browse path; Settings › Your libraries has held the same list, with
// the same removal confirmation, the whole time.

/// While a scan runs.
///
/// `local-library` requires progress reported as a count of items found, and
/// requires that browsing what is already found is not blocked — so this is only
/// ever seen before the first publication arrives.
struct ScanningView: View {
    @Environment(\.theme) private var theme

    let state: LibraryScanState

    var body: some View {
        VStack(spacing: StoryArcSpace.md) {
            ProgressView()
            if case let .scanning(found) = state {
                Text("library.scanning \(found)", bundle: .module)
                    .textRole(.subheadline)
                    .foregroundStyle(theme.palette.textSecondary)
                    .monospacedDigit()
            }
        }
    }
}

/// The first thing a reader ever sees when they own nothing.
///
/// `sources`, the *Adding the first source* scenario, states the whole of it: "one sentence
/// in plain language, one primary action that opens a comic from the device with nothing to
/// configure first, and one plain secondary action that leads to connecting a library", and
/// "the four source types are named only after that secondary action is taken".
///
/// What stood here was the opposite of every clause of that: four rows, one per transport,
/// three of them meaningless to the person reading them, two of them inert because the
/// feature behind them was not built — a taxonomy of protocols on a brand-new reader's very
/// first screen. Apple's onboarding guidance is essential information only, and not forcing
/// setup before the core function; opening a comic is two taps to a readable page and
/// configures nothing.
///
/// The four kinds live one level down, in ``AddSourceMenu``, where choosing between them is
/// the question actually being asked. That menu was written and translated and had no
/// caller anywhere in the app — the audit counted its strings among the nine that shipped
/// unreachable. This is its caller.
///
/// The sentence is `library.empty.*` and Home's is the same one, because Home *is* this
/// state when the library is empty (`home-screen`). It arrived as two near-identical pairs
/// — `home.empty.*` and `library.empty.*`, differing by a word in French and by a clause in
/// English — which is how one situation described twice in a four-language app drifts. One
/// pair now, on both destinations and both platforms.
struct EmptyLibraryView: View {
    /// The primary, and deliberately not a source: a file picker configures nothing and
    /// remembers nothing beyond the copy the app keeps.
    var openComic: () -> Void = {}

    var addFolder: () -> Void = {}
    var addCatalogue: () -> Void = {}
    var addKavita: () -> Void = {}
    var addShare: () -> Void = {}

    var body: some View {
        ContentUnavailableView {
            Label {
                Text("library.empty.title", bundle: .module)
            } icon: {
                Image(systemName: "book.closed")
            }
        } description: {
            Text("library.empty.subtitle", bundle: .module)
        } actions: {
            Button(action: openComic) {
                Text("library.openComic", bundle: .module)
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, StoryArcSpace.xs)
            }
            .buttonStyle(.borderedProminent)

            // Plain, and second: a reader who has just installed a comic app wants to read
            // a comic, and the shelf full of them can wait until they know the app opens
            // one. `sources` calls this "one plain secondary action".
            AddSourceMenu(
                addFolder: addFolder,
                importFile: openComic,
                addCatalogue: addCatalogue,
                addKavita: addKavita,
                addShare: addShare
            )
        }
        .frame(maxWidth: StoryArcSpace.huge * 8)
        .reachableAtEveryTextSize()
    }
}

extension View {
    /// Lets an empty state scroll once the reader's text size makes it taller than the screen.
    ///
    /// `ContentUnavailableView` centres its contents and does not scroll them. At the largest
    /// Dynamic Type size that put the secondary action *behind the tab bar* — reachable by
    /// nobody, on the one screen whose whole job is to be reachable. Caught on a booted
    /// simulator at `accessibility-extra-extra-extra-large`, which is why AGENTS.md §6 asks
    /// for that size by name and why a `#Preview` would not have found it.
    ///
    /// A minimum height rather than a fixed one: at every ordinary size the view still fills
    /// the screen and stays centred, and the scroll view only has somewhere to go when the
    /// content genuinely outgrows it.
    func reachableAtEveryTextSize() -> some View {
        GeometryReader { proxy in
            ScrollView {
                frame(maxWidth: .infinity, minHeight: proxy.size.height)
            }
            .scrollBounceBehavior(.basedOnSize)
        }
    }
}

/// Sources are configured, and the shelf still has nothing on it.
///
/// Two ways to arrive, and they are not the same fact, so they do not get the same
/// sentence: either nothing the reader added can be reached, or the places are answering
/// and have sent nothing to this device yet. Telling a reader on a train that their library
/// is empty would be a lie about their books; telling a reader with a fresh, reachable
/// server that nothing can be reached would be a lie about their network.
///
/// What stood here was ``SourceList`` — the configured sources, their connection states and
/// a coloured dot each. That is the plumbing wearing the shelf's clothes, and §6.2 of the
/// design direction puts connections in Settings and nowhere else on the browse path. They
/// are still there, under Settings › Your libraries, with the same removal flow.
///
/// Never a dead end: one action asks every source again, one opens a comic that needs no
/// source at all. Offline is a normal state, so neither sentence is an error and neither is
/// red — see AGENTS.md §2.
struct LibraryAway: View {
    /// Whether nothing the reader added can be reached.
    ///
    /// Static and pure so the branch can be asserted without a window: which of the two
    /// sentences a reader is shown is the whole substance of this view, and a view is where
    /// a test cannot reach it. A local folder is marked `connected` the moment it is added,
    /// so a configured folder makes this false — which is right, because a folder with
    /// nothing in it has not gone away, it is empty, and those are two different sentences.
    nonisolated static func everythingAway(in registry: SourceRegistry) -> Bool {
        !registry.sources.isEmpty && registry.sources.allSatisfy { !$0.state.canFetch }
    }

    /// Whether *nothing* the reader added can be reached, as opposed to nothing having
    /// arrived yet from what can be.
    let isEverythingAway: Bool

    let retry: () -> Void
    let openComic: () -> Void

    var body: some View {
        ContentUnavailableView {
            Label {
                Text("library.empty.title", bundle: .module)
            } icon: {
                Image(systemName: isEverythingAway ? "wifi.slash" : "books.vertical")
            }
        } description: {
            if isEverythingAway {
                Text("library.away.body", bundle: .module)
            } else {
                Text("library.pending.body", bundle: .module)
            }
        } actions: {
            Button(action: retry) {
                Text("source.offline.retry", bundle: .module)
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, StoryArcSpace.xs)
            }
            .buttonStyle(.borderedProminent)

            Button(action: openComic) {
                Text("library.openComic", bundle: .module)
            }
            .buttonStyle(.plain)
        }
        .frame(maxWidth: StoryArcSpace.huge * 8)
        .reachableAtEveryTextSize()
    }
}

/// A folder that was remembered and can no longer be read.
///
/// `local-library`: "the source is marked `unauthorized` with a plain-language
/// explanation naming the folder", and "a single action re-picks the folder,
/// preserving reading progress for everything inside it". Progress survives
/// because ADR-0006 keys it on the publication, not on the folder.
struct UnavailableFolderNotice: View {
    let name: String
    let repick: () -> Void

    var body: some View {
        HStack(spacing: StoryArcSpace.sm) {
            Text("library.folderUnavailable \(name)", bundle: .module)
                .textRole(.footnote)
                // On glass, so the material decides rather than a fixed palette
                // colour — ``View/storyArcGlassText(_:)`` carries the argument.
                .storyArcGlassText()

            Spacer(minLength: 0)

            Button(action: repick) {
                Text("library.repick", bundle: .module)
                    .textRole(.footnote)
            }
        }
        .padding(.horizontal, StoryArcSpace.md)
        .padding(.vertical, StoryArcSpace.sm)
        // A capsule like the rest of this app's chrome. It keeps its width here, because a
        // sentence with an action beside it needs the room the sentence takes; what it
        // stops doing is running to both edges of the window.
        .storyArcGlass()
        .padding(.horizontal, StoryArcSpace.gutter)
    }
}

File: apps/android/feature/library/src/main/kotlin/app/storyarc/feature/library/LibraryStates.kt
package app.storyarc.feature.library

import android.text.format.DateUtils
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CreateNewFolder
import androidx.compose.material.icons.filled.Dns
import androidx.compose.material.icons.filled.FileDownload
import androidx.compose.material.icons.filled.RssFeed
import androidx.compose.material.icons.filled.Storage
import androidx.compose.material3.Button
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import app.storyarc.core.designsystem.theme.LocalStoryArcPalette
import app.storyarc.core.designsystem.tokens.StoryArcSpace
import app.storyarc.core.model.SourceRegistry

/**
 * States that the shelf on screen is last session's, and when it was confirmed.
 *
 * `sources` asks for "a single unobtrusive indicator" saying "that content is cached and
 * when it was last refreshed". One line, in the secondary text colour, that leaves as soon
 * as a walk finishes — at which point the shelf is not cached, it is current, and a notice
 * still claiming otherwise would be the indicator lying quietly in the corner.
 *
 * Not an error and not a warning. Offline is a normal state; so is a library that has not
 * been rewalked yet. iOS shows the same line above its grid, from `CachedNotice`.
 */
@Composable
internal fun CachedNotice(refreshedAtEpochMillis: Long) {
    val palette = LocalStoryArcPalette.current
    // The platform's own phrasing for "twelve minutes ago", which `localization` requires
    // rather than a string this app assembles and then has to translate four times.
    val relative = DateUtils.getRelativeTimeSpanString(
        refreshedAtEpochMillis,
        System.currentTimeMillis(),
        DateUtils.MINUTE_IN_MILLIS,
    ).toString()

    Text(
        text = stringResource(R.string.library_cached, relative),
        style = MaterialTheme.typography.labelLarge,
        color = palette.textSecondary,
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = StoryArcSpace.gutter, vertical = StoryArcSpace.xs),
    )
}

/**
 * The first thing a reader ever sees when they own nothing.
 *
 * `sources`, the *Adding the first source* scenario, states the whole of it: "one sentence
 * in plain language, one primary action that opens a comic from the device with nothing to
 * configure first, and one plain secondary action that leads to connecting a library", and
 * "the four source types are named only after that secondary action is taken".
 *
 * What stood here — in `LibraryScreen.kt`, which is over the module's line cap and is why
 * this moved — was the opposite of every clause of that: four rows, one per transport,
 * three of them meaningless to the person reading them, over a primary button that read
 * *Refresh the library* while actually opening the folder picker, because one string served
 * both it and the toolbar's refresh icon.
 *
 * Hand-composed rather than a port of iOS's `ContentUnavailableView`, per divergence #12:
 * Material publishes no empty-state component, and importing iOS's shape is the
 * cross-platform habit this revamp is undoing. The *content model* is the same, and so are
 * the words.
 */
@Composable
internal fun EmptyLibrary(
    /** Copies one file in and opens it. Configures nothing, remembers nothing else. */
    onOpenComic: () -> Unit,
    onAddFolder: () -> Unit,
    onAddCatalogue: () -> Unit,
    onAddKavita: () -> Unit,
    onAddShare: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val palette = LocalStoryArcPalette.current

    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = StoryArcSpace.gutter, vertical = StoryArcSpace.xxl),
        verticalArrangement = Arrangement.spacedBy(StoryArcSpace.md),
    ) {
        Text(
            text = stringResource(R.string.library_empty_title),
            style = MaterialTheme.typography.headlineSmall,
            color = palette.textPrimary,
        )
        Text(
            text = stringResource(R.string.library_empty_subtitle),
            style = MaterialTheme.typography.bodyLarge,
            color = palette.textSecondary,
            modifier = Modifier
                .widthIn(max = FIRST_RUN_MEASURE)
                .padding(bottom = StoryArcSpace.sm),
        )
        Button(onClick = onOpenComic) {
            Text(stringResource(R.string.library_open_comic))
        }
        // Plain, and second: a reader who has just installed a comic app wants to read a
        // comic, and the shelf full of them can wait until they know the app opens one.
        AddBooksButton(
            onAddFolder = onAddFolder,
            onOpenComic = onOpenComic,
            onAddCatalogue = onAddCatalogue,
            onAddKavita = onAddKavita,
            onAddShare = onAddShare,
        )
    }
}

/**
 * Sources are configured, and the shelf still has nothing on it.
 *
 * Two ways to arrive, and they are not the same fact, so they do not get the same sentence:
 * either nothing the reader added can be reached, or the places are answering and have sent
 * nothing to this device yet. Telling a reader on a train that their library is empty would
 * be a lie about their books; telling a reader with a fresh, reachable server that nothing
 * can be reached would be a lie about their network.
 *
 * What stood here was the source list — the configured sources, their connection states and
 * a coloured dot each. That is the plumbing wearing the shelf's clothes, and §6.2 of the
 * design direction puts connections in Settings and nowhere else on the browse path. They
 * are still there, with the same removal flow.
 *
 * Never a dead end: one action asks every source again, one opens a comic that needs no
 * source at all. Offline is a normal state, so neither sentence is an error and neither is
 * red — see AGENTS.md §2. iOS shows the same pair from `LibraryAway`.
 */
@Composable
internal fun LibraryAway(
    isEverythingAway: Boolean,
    onRetry: () -> Unit,
    onOpenComic: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val palette = LocalStoryArcPalette.current

    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = StoryArcSpace.gutter, vertical = StoryArcSpace.xxl),
        verticalArrangement = Arrangement.spacedBy(StoryArcSpace.md),
    ) {
        Text(
            text = stringResource(R.string.library_empty_title),
            style = MaterialTheme.typography.headlineSmall,
            color = palette.textPrimary,
        )
        Text(
            text = stringResource(
                if (isEverythingAway) R.string.library_away_body else R.string.library_pending_body,
            ),
            style = MaterialTheme.typography.bodyLarge,
            color = palette.textSecondary,
            modifier = Modifier
                .widthIn(max = FIRST_RUN_MEASURE)
                .padding(bottom = StoryArcSpace.sm),
        )
        Button(onClick = onRetry) {
            Text(stringResource(R.string.source_offline_retry))
        }
        TextButton(onClick = onOpenComic) {
            Text(stringResource(R.string.library_open_comic))
        }
    }
}

/**
 * The four kinds of place, one level down.
 *
 * A labelled button rather than the toolbar's plus glyph: on an empty screen the reader has
 * nothing to compare a lone icon against, and `sources` asks for a *plain secondary action*,
 * which is a thing with words on it. What it opens is the same five choices the toolbar
 * offers, and the four transports are named only here — where choosing between them is the
 * question actually being asked, rather than a wall to be understood first.
 */
@Composable
private fun AddBooksButton(
    onAddFolder: () -> Unit,
    onOpenComic: () -> Unit,
    onAddCatalogue: () -> Unit,
    onAddKavita: () -> Unit,
    onAddShare: () -> Unit,
) {
    var open by remember { mutableStateOf(false) }

    TextButton(onClick = { open = true }) {
        Text(stringResource(R.string.library_add_source))
    }
    DropdownMenu(expanded = open, onDismissRequest = { open = false }) {
        AddBooksItem(R.string.library_add_folder, Icons.Filled.CreateNewFolder) {
            open = false
            onAddFolder()
        }
        AddBooksItem(R.string.library_import, Icons.Filled.FileDownload) {
            open = false
            onOpenComic()
        }
        AddBooksItem(R.string.catalogue_title, Icons.Filled.RssFeed) {
            open = false
            onAddCatalogue()
        }
        AddBooksItem(R.string.kavita_title, Icons.Filled.Dns) {
            open = false
            onAddKavita()
        }
        AddBooksItem(R.string.smb_title, Icons.Filled.Storage) {
            open = false
            onAddShare()
        }
    }
}

@Composable
private fun AddBooksItem(
    label: Int,
    icon: androidx.compose.ui.graphics.vector.ImageVector,
    onClick: () -> Unit,
) {
    DropdownMenuItem(
        text = { Text(stringResource(label)) },
        leadingIcon = { Icon(icon, contentDescription = null) },
        onClick = onClick,
    )
}

/**
 * Whether nothing the reader added can be reached.
 *
 * A function rather than a line inside the composable so the branch can be asserted without
 * a device: which of [LibraryAway]'s two sentences a reader is shown is the whole substance
 * of that screen. A local folder is marked connected the moment it is added, so a
 * configured folder makes this false — which is right, because a folder with nothing in it
 * has not gone away, it is empty, and those are two different sentences.
 *
 * iOS answers the same question in `LibraryAway.everythingAway(in:)`.
 */
internal fun everythingAway(registry: SourceRegistry): Boolean =
    registry.sources.isNotEmpty() && registry.sources.none { it.state.canFetch }

/**
 * How wide a first-run sentence is allowed to run.
 *
 * A tablet would otherwise set one line across the whole window, which is the measure this
 * app spends the rest of its effort avoiding.
 */
private val FIRST_RUN_MEASURE = 520.dp
